### PR TITLE
feat(rag): RAG-in-a-box CLI (rostam-server rag ingest/ask/query)

### PR DESCRIPTION
**Pull request by vahid-sohrabloo**
## RAG-in-a-box CLI

A single-binary retrieval-augmented-generation CLI over the existing Rostam engine:

```
rostam-server rag ingest ./docs
rostam-server rag ask "how does the epoll transport pick its loop count?"
rostam-server rag query "epoll loop count"   # retrieval only, no LLM
```

`ask` retrieves the most relevant chunks and calls an LLM to synthesize a grounded, cited answer. `query` returns ranked chunks (no LLM). Built as roadmap **option C** on top of the merged MCP server (#12) and LLM proxy (#14).

### Design (Approach A — thin orchestrator over the existing engine)
- New `rag/` package: `chunk` → `store` (a `Retriever` interface with an in-process embedded backend and an HTTP `--endpoint` backend) → `ingest` → `retrieve` → `ask`. All engine coupling is isolated in `store.go`; the rest is backend-agnostic.
- Reuses `vector.CollectionStore` (`SearchDocs`/`SearchText`/`Upsert`/`DeleteByFilter`), the `semcache` embedder abstraction, and a small self-contained OpenAI-compatible chat client (no `llmproxy` coupling).
- **Embedded by default** (local `--data` dir, nothing to run); `--endpoint` targets a running server.
- **Retrieval:** BM25 when no embedder is configured; dense KNN when one is. (Dense+BM25 *fusion* is an intentional fast-follow — see below.)
- **Secrets are env-only** (`ROSTAM_EMBED_KEY` / `ROSTAM_LLM_KEY`), never flags, matching the repo convention.
- Ingest is **idempotent per source file** (delete-then-upsert with deterministic chunk IDs); re-ingesting an emptied file purges its stale chunks; a dim-change re-ingest is refused with an actionable message rather than a raw engine error.

### Scope
- 17 files, +1594 lines. Blast radius outside the new package/docs is **4 lines** in `cmd/rostam-server/main.go` (one dispatch clause matching the existing `mcp`/`llm-proxy` shape).
- Docs: `docs/server/rag.md` (+ mkdocs nav) and a README section.

### Testing
- `go build ./...`, `go vet`, and `go test ./rag/ ./cmd/rostam-server/` all green.
- Backend parity is covered by a real over-the-wire test that stands up an in-process `rostam.NewServer` and runs the same assertions as the embedded backend (not just a compile-time interface check).

### Fast-follow (not blockers, tracked)
- Dense+BM25 fusion (`HybridText` returns content-less results — needs a content-by-id fetch).
- HTTP-side `EnsureCorpus` dim guard (the client has no read-collection-dim RPC today; embedded path guards, HTTP surfaces the raw error).
- `HTTPRetriever` discards `FanMeta` (partial fan-out signal — only relevant to partitioned collections via `--endpoint`).
- Minor polish: structured-slog shape in one `fatal`, include the LLM error-response body on non-200, a couple of extra edge-case tests.

## Summary by CodeRabbit

- **New Features**
  - Added a RAG command for document ingestion, passage retrieval, and cited question answering.
  - Supports BM25 and dense retrieval, local or remote operation, configurable chunking, embeddings, and LLM providers.
  - Added recursive ingestion for supported text files with idempotent re-ingestion.
  - Added local and HTTP-backed retrieval options.

- **Documentation**
  - Added comprehensive RAG CLI documentation, configuration guidance, supported file types, and offline usage instructions.

### CODE REVIEW

File: rag/chunk.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+// SplitText splits s into overlapping chunks of about targetTokens words each,
+// preferring blank-line paragraph boundaries. See the interface block above.
+func SplitText(s string, targetTokens, overlap int) []Chunk {
+	if targetTokens <= 0 {
+		targetTokens = 512
+	}
+	if overlap < 0 || overlap >= targetTokens {
+		overlap = 0
+	}
```
[the remark this answers is not in the document]
Comment (vahid-sohrabloo) [reply rag/chunk.go:24]: Not applicable: the CLI passes `chunk-overlap=0`, but `Ingest` coerces it to the documented 64 before calling `SplitText` (rag/ingest.go: `if opts.ChunkOverlap <= 0 { opts.ChunkOverlap = 64 }`, and ChunkSize→512 likewise). So default ingestion uses 64-word overlap, not disjoint chunks — the flag help even reads "0 = rag package default, 64". `SplitText` leaves 0 unchanged in isolation, but no CLI path reaches it with 0. Resolving.

File: rag/ingest.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+			}
+			if d.IsDir() {
+				return nil
+			}
+			if !textExts[strings.ToLower(filepath.Ext(path))] {
+				rep.Skipped++
+				rep.SkippedPaths = append(rep.SkippedPaths, path)
+				return nil
+			}
+			body, err := os.ReadFile(path)
```
[the remark this answers is not in the document]
Comment (vahid-sohrabloo) [reply rag/ingest.go]: Addressed in 0c84316: added a justified `//nolint:gosec // G122: ...` on that `os.ReadFile`, matching the repo's established convention for justified gosec sites (sysmem_linux.go, cache/page.go, etc.). The ingest paths are the user's own local files for a CLI; symlink TOCTOU is not in the threat model here. golangci-lint is green.

File: rag/ask.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	hits, err := Retrieve(ctx, r, emb, corpus, question, k)
+	if err != nil {
+		return AskResult{}, err
+	}
+	sys, user := BuildPrompt(question, hits)
+	answer, err := chatComplete(ctx, llm, sys, user)
+	if err != nil {
+		return AskResult{}, err
+	}
+	return AskResult{Answer: answer, Hits: hits}, nil
```
[the remark this answers is not in the document]
Comment (vahid-sohrabloo) [reply rag/ask.go:62]: Deferring to fast-follow (tracked in the PR description). Enforcing "answer must contain >=1 citation or error" risks false rejections — an LLM can produce a correctly grounded answer without emitting a [n] token, and the explicit no-answer path legitimately has none. Rejecting out-of-range indexes is cheap and safe, but the overall cited-answer contract is a behavioral-policy decision worth deliberate design (what counts as substantive, how to parse [n], grounded-but-uncited handling) rather than a quick gate bolted on at merge. Not a blocker for this MVP.

File: rag/ingest.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+			Content: c.Content,
+			Source:  source,
+			Index:   c.Index,
+		}
+		if vecs != nil {
+			sc.Vector = vecs[i]
+		}
+		stored[i] = sc
+	}
+	if err := r.Upsert(ctx, opts.Corpus, stored); err != nil {
```
[the remark this answers is not in the document]
Comment (vahid-sohrabloo) [reply rag/ingest.go:136]: Deferring to fast-follow (in the PR description). This is the transient-loss-on-failed-re-embed case the whole-branch review already assessed: the loss is not permanent (chunk IDs are deterministic in source+index, so re-running after the embedder recovers restores the content), and failing rather than half-indexing is the intended stance. An atomic staged-generation replace is a Retriever-API change worth doing deliberately, not in this PR.
